File: services/routeService.js
import api from './apiClient';
import { BASE_URL, API_ENDPOINTS } from '../constants/config';

/**
 * Get driver's routes/trips by status filter
 * @param {object} params - { statusFilter, bookingDate }
 * @param {string} params.statusFilter - 'upcoming'|'ongoing'|'completed'
 * @param {string} params.bookingDate - YYYY-MM-DD format
 * @returns {Promise<{success: boolean, routes?: array, count?: number, error?: string}>}
 */
export async function getDriverTrips({ statusFilter = 'upcoming', bookingDate = null }) {
  try {
    // Format date as YYYY-MM-DD if not provided
    let dateParam = bookingDate;
    if (!dateParam) {
      const today = new Date();
      const year = today.getFullYear();
      const month = String(today.getMonth() + 1).padStart(2, '0');
      const day = String(today.getDate()).padStart(2, '0');
      dateParam = `${year}-${month}-${day}`;
    }

    const url = `${BASE_URL}${API_ENDPOINTS.DRIVER_TRIPS}?status_filter=${statusFilter}&booking_date=${dateParam}`;
    
    console.log('=== Fetching Driver Trips ===');
    console.log('URL:', url);
    console.log('Status Filter:', statusFilter);
    console.log('Booking Date:', dateParam);
    
    const res = await api.get(url);
    
    console.log('Driver Trips Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success && res.data.data) {
      const routes = res.data.data.routes || [];
      const count = res.data.data.count || 0;
      
      console.log('Routes fetched:', count);
      
      return { 
        success: true, 
        routes: routes,
        count: count,
        message: res.data.message
      };
    } else {
      return { success: true, routes: [], count: 0 };
    }
  } catch (err) {
    console.log('=== Driver Trips Error ===');
    console.log('Error:', err.message);
    
    if (err.response) {
      console.log('Response status:', err.response.status);
      console.log('Response data:', JSON.stringify(err.response.data, null, 2));
      
      let errorMessage = 'Failed to fetch routes';
      
      if (err.response.data) {
        const data = err.response.data;
        
        if (data.detail && typeof data.detail.message === 'string') {
          errorMessage = data.detail.message;
        } else if (typeof data.message === 'string') {
          errorMessage = data.message;
        } else if (typeof data.error === 'string') {
          errorMessage = data.error;
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        }
      }
      
      return { success: false, error: errorMessage };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * Start driver's duty on a route (v2.0)
 * Changes route status from DRIVER_ASSIGNED → ONGOING
 * @param {number} routeId - Route ID to start duty on
 * @returns {Promise<{success: boolean, data?: object, error?: string}>}
 */
export async function startDuty(routeId) {
  try {
    const url = `${BASE_URL}${API_ENDPOINTS.DUTY_START}?route_id=${routeId}`;
    
    console.log('=== Starting Duty ===');
    console.log('Route ID:', routeId);
    console.log('URL:', url);
    
    const res = await api.post(url);
    
    console.log('Start Duty Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success) {
      return {
        success: true,
        data: res.data.data,
        message: res.data.message
      };
    }
    
    return { success: false, error: 'Failed to start duty' };
  } catch (err) {
    console.log('=== Start Duty Error ===');
    console.log('Error:', err.message);
    
    if (err.response && err.response.data) {
      const data = err.response.data;
      console.log('Error Response Data:', JSON.stringify(data, null, 2));
      
      let errorMessage = 'Failed to start duty';
      
      // Handle FastAPI validation errors (422 status)
      if (err.response.status === 422 && data.detail) {
        if (Array.isArray(data.detail)) {
          errorMessage = data.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const message = err.msg || 'Invalid value';
            return `${field}: ${message}`;
          }).join(', ');
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        } else if (typeof data.detail === 'object') {
          errorMessage = data.detail.msg || data.detail.message || JSON.stringify(data.detail);
        }
      } else {
        errorMessage = data.message || data.detail || data.error || 'Failed to start duty';
      }
      
      let errorCode = data.error_code || null;
      let details = data.details || null;
      
      return { 
        success: false, 
        error: errorMessage,
        errorCode: errorCode,
        details: details
      };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * End driver's duty and complete the route (v2.0)
 * Changes route status from ONGOING → COMPLETED
 * @param {number} routeId - Route ID to end duty on
 * @param {string} reason - Optional reason for early termination
 * @returns {Promise<{success: boolean, data?: object, error?: string}>}
 */
export async function endDuty(routeId, reason = null) {
  try {
    let url = `${BASE_URL}${API_ENDPOINTS.DUTY_END}?route_id=${routeId}`;
    if (reason) {
      url += `&reason=${encodeURIComponent(reason)}`;
    }
    
    console.log('=== Ending Duty ===');
    console.log('Route ID:', routeId);
    console.log('URL:', url);
    
    const res = await api.put(url);
    
    console.log('End Duty Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success) {
      return {
        success: true,
        data: res.data.data,
        message: res.data.message
      };
    }
    
    return { success: false, error: 'Failed to end duty' };
  } catch (err) {
    console.log('=== End Duty Error ===');
    console.log('Error:', err.message);
    
    if (err.response && err.response.data) {
      const data = err.response.data;
      console.log('Error Response Data:', JSON.stringify(data, null, 2));
      
      let errorMessage = 'Failed to end duty';
      let errorCode = null;
      let details = null;
      
      // Handle 400 errors with detail object (business logic errors)
      if (err.response.status === 400 && data.detail && typeof data.detail === 'object') {
        errorMessage = data.detail.message || 'Request failed';
        errorCode = data.detail.error_code || null;
        details = data.detail.details || null;
      }
      // Handle FastAPI validation errors (422 status)
      else if (err.response.status === 422 && data.detail) {
        if (Array.isArray(data.detail)) {
          errorMessage = data.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const message = err.msg || 'Invalid value';
            return `${field}: ${message}`;
          }).join(', ');
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        } else if (typeof data.detail === 'object') {
          errorMessage = data.detail.msg || data.detail.message || JSON.stringify(data.detail);
        }
      }
      // Handle 404 errors
      else if (err.response.status === 404 && data.detail) {
        if (typeof data.detail === 'object') {
          errorMessage = data.detail.message || 'Not found';
          errorCode = data.detail.error_code || null;
        } else {
          errorMessage = data.detail;
        }
      }
      // Handle other error formats
      else {
        errorMessage = data.message || data.detail || data.error || 'Failed to end duty';
      }
      
      return { 
        success: false, 
        error: errorMessage,
        errorCode: errorCode,
        details: details
      };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * Start trip - Mark passenger pickup (v2.0)
 * Changes booking status from REQUEST/SCHEDULED → ONGOING
 * @param {object} params - { routeId, bookingId, otp, latitude, longitude }
 * @returns {Promise<{success: boolean, data?: object, error?: string}>}
 */
export async function startTrip({ routeId, bookingId, otp, latitude, longitude }) {
  try {
    let url = `${BASE_URL}${API_ENDPOINTS.TRIP_START}?route_id=${routeId}&booking_id=${bookingId}`;
    
    if (otp) {
      url += `&otp=${otp}`;
    }
    if (latitude !== undefined && longitude !== undefined) {
      url += `&current_latitude=${latitude}&current_longitude=${longitude}`;
    }
    
    console.log('=== Starting Trip (Pickup) ===');
    console.log('Route ID:', routeId);
    console.log('Booking ID:', bookingId);
    console.log('URL:', url);
    
    const res = await api.post(url);
    
    console.log('Start Trip Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success) {
      return {
        success: true,
        data: res.data.data,
        message: res.data.message
      };
    }
    
    return { success: false, error: 'Failed to start trip' };
  } catch (err) {
    console.log('=== Start Trip Error ===');
    console.log('Error:', err.message);
    
    if (err.response && err.response.data) {
      const data = err.response.data;
      console.log('Error Response Data:', JSON.stringify(data, null, 2));
      
      let errorMessage = 'Failed to start trip';
      let errorCode = null;
      let details = null;
      
      // Handle 400 errors with detail object (business logic errors)
      if (err.response.status === 400 && data.detail && typeof data.detail === 'object') {
        errorMessage = data.detail.message || 'Request failed';
        errorCode = data.detail.error_code || null;
        details = data.detail.details || null;
      }
      // Handle FastAPI validation errors (422 status)
      else if (err.response.status === 422 && data.detail) {
        if (Array.isArray(data.detail)) {
          // Validation errors array
          errorMessage = data.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const message = err.msg || 'Invalid value';
            return `${field}: ${message}`;
          }).join(', ');
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        } else if (typeof data.detail === 'object') {
          errorMessage = data.detail.msg || data.detail.message || JSON.stringify(data.detail);
        }
      }
      // Handle 404 errors
      else if (err.response.status === 404 && data.detail) {
        if (typeof data.detail === 'object') {
          errorMessage = data.detail.message || 'Not found';
          errorCode = data.detail.error_code || null;
        } else {
          errorMessage = data.detail;
        }
      }
      // Handle other error formats
      else {
        errorMessage = data.message || data.detail || data.error || 'Failed to start trip';
      }
      
      return { 
        success: false, 
        error: errorMessage,
        errorCode: errorCode,
        details: details
      };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * Drop passenger (v2.0)
 * Changes booking status from ONGOING → COMPLETED
 * @param {object} params - { routeId, bookingId, otp, latitude, longitude }
 * @returns {Promise<{success: boolean, data?: object, error?: string}>}
 */
export async function dropTrip({ routeId, bookingId, otp, latitude, longitude }) {
  try {
    let url = `${BASE_URL}${API_ENDPOINTS.TRIP_DROP}?route_id=${routeId}&booking_id=${bookingId}`;
    
    if (otp) {
      url += `&otp=${otp}`;
    }
    if (latitude !== undefined && longitude !== undefined) {
      url += `&current_latitude=${latitude}&current_longitude=${longitude}`;
    }
    
    console.log('=== Dropping Passenger ===');
    console.log('Route ID:', routeId);
    console.log('Booking ID:', bookingId);
    console.log('URL:', url);
    
    const res = await api.put(url);
    
    console.log('Drop Trip Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success) {
      return {
        success: true,
        data: res.data.data,
        message: res.data.message
      };
    }
    
    return { success: false, error: 'Failed to drop passenger' };
  } catch (err) {
    console.log('=== Drop Trip Error ===');
    console.log('Error:', err.message);
    
    if (err.response && err.response.data) {
      const data = err.response.data;
      console.log('Error Response Data:', JSON.stringify(data, null, 2));
      
      let errorMessage = 'Failed to drop passenger';
      let errorCode = null;
      let details = null;
      
      // Handle 400 errors with detail object (business logic errors)
      if (err.response.status === 400 && data.detail && typeof data.detail === 'object') {
        errorMessage = data.detail.message || 'Request failed';
        errorCode = data.detail.error_code || null;
        details = data.detail.details || null;
      }
      // Handle FastAPI validation errors (422 status)
      else if (err.response.status === 422 && data.detail) {
        if (Array.isArray(data.detail)) {
          errorMessage = data.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const message = err.msg || 'Invalid value';
            return `${field}: ${message}`;
          }).join(', ');
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        } else if (typeof data.detail === 'object') {
          errorMessage = data.detail.msg || data.detail.message || JSON.stringify(data.detail);
        }
      }
      // Handle 404 errors
      else if (err.response.status === 404 && data.detail) {
        if (typeof data.detail === 'object') {
          errorMessage = data.detail.message || 'Not found';
          errorCode = data.detail.error_code || null;
        } else {
          errorMessage = data.detail;
        }
      }
      // Handle other error formats
      else {
        errorMessage = data.message || data.detail || data.error || 'Failed to drop passenger';
      }
      
      return { 
        success: false, 
        error: errorMessage,
        errorCode: errorCode,
        details: details
      };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * Mark passenger as no-show (v2.0)
 * Changes booking status to NO_SHOW
 * @param {object} params - { routeId, bookingId, reason }
 * @returns {Promise<{success: boolean, data?: object, error?: string}>}
 */
export async function markNoShow({ routeId, bookingId, reason }) {
  try {
    let url = `${BASE_URL}${API_ENDPOINTS.TRIP_NO_SHOW}?route_id=${routeId}&booking_id=${bookingId}`;
    
    if (reason) {
      url += `&reason=${encodeURIComponent(reason)}`;
    }
    
    console.log('=== Marking No-Show ===');
    console.log('Route ID:', routeId);
    console.log('Booking ID:', bookingId);
    console.log('Reason:', reason);
    console.log('URL:', url);
    
    const res = await api.put(url);
    
    console.log('No-Show Response:', JSON.stringify(res.data, null, 2));
    
    if (res.data && res.data.success) {
      return {
        success: true,
        data: res.data.data,
        message: res.data.message
      };
    }
    
    return { success: false, error: 'Failed to mark no-show' };
  } catch (err) {
    console.log('=== No-Show Error ===');
    console.log('Error:', err.message);
    
    if (err.response && err.response.data) {
      const data = err.response.data;
      console.log('Error Response Data:', JSON.stringify(data, null, 2));
      
      let errorMessage = 'Failed to mark no-show';
      let errorCode = null;
      let details = null;
      
      // Handle 400 errors with detail object (business logic errors)
      if (err.response.status === 400 && data.detail && typeof data.detail === 'object') {
        errorMessage = data.detail.message || 'Request failed';
        errorCode = data.detail.error_code || null;
        details = data.detail.details || null;
      }
      // Handle FastAPI validation errors (422 status)
      else if (err.response.status === 422 && data.detail) {
        if (Array.isArray(data.detail)) {
          errorMessage = data.detail.map(err => {
            const field = err.loc ? err.loc.join('.') : 'field';
            const message = err.msg || 'Invalid value';
            return `${field}: ${message}`;
          }).join(', ');
        } else if (typeof data.detail === 'string') {
          errorMessage = data.detail;
        } else if (typeof data.detail === 'object') {
          errorMessage = data.detail.msg || data.detail.message || JSON.stringify(data.detail);
        }
      }
      // Handle 404 errors
      else if (err.response.status === 404 && data.detail) {
        if (typeof data.detail === 'object') {
          errorMessage = data.detail.message || 'Not found';
          errorCode = data.detail.error_code || null;
        } else {
          errorMessage = data.detail;
        }
      }
      // Handle other error formats
      else {
        errorMessage = data.message || data.detail || data.error || 'Failed to mark no-show';
      }
      
      return { 
        success: false, 
        error: errorMessage,
        errorCode: errorCode,
        details: details
      };
    }
    
    return { success: false, error: 'Network error. Please try again.' };
  }
}

/**
 * Helper: Set active duty status in local storage
 */
export async function setActiveDutyStatus() {
  try {
    const AsyncStorage = (await import('@react-native-async-storage/async-storage')).default;
    await AsyncStorage.setItem('has_active_duty', 'true');
    console.log('✅ Active duty status set');
  } catch (error) {
    console.log('Error setting duty status:', error);
  }
}

/**
 * Helper: Clear active duty status from local storage
 */
export async function clearActiveDutyStatus() {
  try {
    const AsyncStorage = (await import('@react-native-async-storage/async-storage')).default;
    await AsyncStorage.removeItem('has_active_duty');
    console.log('✅ Active duty status cleared');
  } catch (error) {
    console.log('Error clearing duty status:', error);
  }
}

/**
 * Helper: Check if driver has active duty
 */
export async function checkActiveDutyStatus() {
  try {
    const AsyncStorage = (await import('@react-native-async-storage/async-storage')).default;
    const status = await AsyncStorage.getItem('has_active_duty');
    return status === 'true';
  } catch (error) {
    console.log('Error checking duty status:', error);
    return false;
  }
}

File: lib/services/route_service.dart
import 'package:dio/dio.dart';
import 'package:logger/logger.dart';
import '../config/constants.dart';
import 'api_client.dart';

class RouteService {
  final Dio _dio = ApiClient().client;
  final Logger _logger = Logger();

  // Singleton
  static final RouteService _instance = RouteService._internal();
  factory RouteService() => _instance;
  RouteService._internal();

  /// Fetch Driver Trips
  Future<Map<String, dynamic>> getDriverTrips({String statusFilter = 'upcoming', String? bookingDate}) async {
    try {
      final queryParams = <String, dynamic>{
        'status_filter': statusFilter,
      };

      if (bookingDate != null) {
        queryParams['booking_date'] = bookingDate;
      } else if (statusFilter == 'upcoming') {
        final now = DateTime.now();
        final formatted = "${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}";
        queryParams['booking_date'] = formatted;
      }

      _logger.i('Fetching trips: $queryParams');

      final response = await _dio.get(ApiEndpoints.driverTrips, queryParameters: queryParams);
      
      // RN: result.routes = res.data.data.routes
      final data = response.data['data'];
      final routes = (data is Map<String, dynamic> && data.containsKey('routes')) 
          ? data['routes'] 
          : (data is List ? data : []); // Fallback if API changes
          
      return {
        'success': true,
        'routes': routes is List ? routes : []
      };

    } on DioException catch (e) {
      _logger.e('Failed to fetch trips', error: e);
      return _handleError(e);
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Fetch Employee Bookings (My Schedules for Employee/Driver)
  Future<Map<String, dynamic>> getEmployeeBookings({required Map<String, dynamic> params}) async {
    try {
      _logger.i('Fetching bookings: $params');
      final response = await _dio.get(ApiEndpoints.bookings, queryParameters: params);
      
      final data = response.data['data']; // Assuming standard response structure
      // RN app expects result.bookings
      return {
        'success': true,
        'bookings': data is List ? data : (data['bookings'] is List ? data['bookings'] : [])
      };
    } on DioException catch (e) {
      _logger.e('Failed to fetch bookings', error: e);
      return {'success': false, 'error': e.message};
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Start Duty
  Future<Map<String, dynamic>> startDuty(String routeId) async {
    try {
      _logger.i('Starting duty for route: $routeId');
      // Fix: Use query params, not body
      final response = await _dio.post(
        ApiEndpoints.dutyStart,
        queryParameters: {'route_id': routeId},
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// End Duty
  Future<Map<String, dynamic>> endDuty(String routeId, String? reason) async {
    try {
      _logger.i('Ending duty for route: $routeId');
      
      final params = {'route_id': routeId};
      if (reason != null) params['reason'] = reason;

      // Fix: Use PUT and query params
      final response = await _dio.put(
        ApiEndpoints.dutyEnd,
        queryParameters: params,
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Start Pickup (Start Trip)
  /// [idempotencyKey] — stable client-generated key; safe to retry with the
  ///   same key on network timeout without creating a duplicate stop event.
  /// [deviceTimestamp] — device-local event time as ISO-8601 UTC string.
  Future<Map<String, dynamic>> startTrip({
    required String routeId,
    required String bookingId,
    String? otp,
    required double latitude,
    required double longitude,
    String? idempotencyKey,
    String? deviceTimestamp,
  }) async {
    try {
      _logger.i('Starting trip (Pickup) for booking: $bookingId');
      
      final params = <String, dynamic>{
        'route_id': routeId,
        'booking_id': bookingId,
        'current_latitude': latitude,
        'current_longitude': longitude,
      };
      if (otp != null) params['otp'] = otp;
      if (idempotencyKey != null) params['idempotency_key'] = idempotencyKey;
      if (deviceTimestamp != null) params['device_timestamp'] = deviceTimestamp;

      final response = await _dio.post(
        ApiEndpoints.tripStart,
        queryParameters: params,
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Drop Passenger (End Trip)
  /// [idempotencyKey] — stable client-generated key; safe to retry with the
  ///   same key on network timeout without creating a duplicate stop event.
  /// [deviceTimestamp] — device-local event time as ISO-8601 UTC string.
  Future<Map<String, dynamic>> dropTrip({
    required String routeId,
    required String bookingId,
    String? otp,
    required double latitude,
    required double longitude,
    String? idempotencyKey,
    String? deviceTimestamp,
  }) async {
    try {
      _logger.i('Dropping passenger for booking: $bookingId');
      
      final params = <String, dynamic>{
        'route_id': routeId,
        'booking_id': bookingId,
        'current_latitude': latitude,
        'current_longitude': longitude,
      };
      if (otp != null) params['otp'] = otp;
      if (idempotencyKey != null) params['idempotency_key'] = idempotencyKey;
      if (deviceTimestamp != null) params['device_timestamp'] = deviceTimestamp;

      final response = await _dio.put(
        ApiEndpoints.tripEnd,
        queryParameters: params,
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Mark No Show
  Future<Map<String, dynamic>> markNoShow({
    required String routeId,
    required String bookingId,
    String? reason,
  }) async {
    try {
      _logger.i('Marking No Show for booking: $bookingId');
      
      final params = {
        'route_id': routeId,
        'booking_id': bookingId,
        'reason': reason ?? 'Passenger did not show up',
      };

      // Fix: Use PUT and query params
      final response = await _dio.put(
        ApiEndpoints.noShow,
        queryParameters: params,
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Send a GPS location ping for an ONGOING route — POST /driver/location.
  /// Called every ~7 s while duty is active.
  /// [speedKmh] is optional but should be passed whenever the device GPS
  /// provides it; the server uses it for ETA recalc and speed enforcement.
  /// Throws on non-2xx so the caller can apply retry logic.
  Future<void> sendLocation({
    required String routeId,
    required double latitude,
    required double longitude,
    double? speedKmh,
  }) async {
    final params = <String, dynamic>{
      'route_id': routeId,
      'latitude': latitude,
      'longitude': longitude,
    };
    if (speedKmh != null) params['speed'] = speedKmh;
    await _dio.post(ApiEndpoints.driverLocation, queryParameters: params);
  }

  /// Board the escort — POST /driver/escort/board.
  /// [code] is omitted for `off` mode (board immediately) and required for
  /// `universal` / `unique` modes.
  Future<Map<String, dynamic>> escortBoard({
    required String routeId,
    String? code,
  }) async {
    try {
      _logger.i('Boarding escort for route: $routeId');
      final params = <String, dynamic>{'route_id': routeId};
      if (code != null) params['code'] = code;
      final response = await _dio.post(
        ApiEndpoints.escortBoard,
        queryParameters: params,
      );
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Cancel Booking
  Future<Map<String, dynamic>> cancelBooking(String bookingId) async {
    try {
      _logger.i('Cancelling booking: $bookingId');
      final response = await _dio.post(ApiEndpoints.cancelBooking, data: {
        'booking_id': bookingId,
      });
      return {'success': true, 'data': response.data};
    } on DioException catch (e) {
      return _handleError(e);
    }
  }

  /// Fetch Driver History Report
  Future<Map<String, dynamic>> getDriverHistoryReport({
    required String startDate,
    required String endDate,
  }) async {
    try {
      final queryParams = {
        'start_date': startDate,
        'end_date': endDate,
      };

      _logger.i('Fetching history report: $queryParams');
      final response = await _dio.get(
        ApiEndpoints.driverHistoryReport,
        queryParameters: queryParams,
      );

      final responseData = response.data;
      if (responseData is Map && responseData['success'] == true) {
        final data = responseData['data'] ?? {};
        return {
          'success': true,
          'summary': data['summary'] ?? {},
          'bookings': data['bookings'] ?? [],
        };
      }

      return {
        'success': false,
        'error': (responseData is Map ? responseData['message'] : null) ?? 'Failed to load report',
      };
    } on DioException catch (e) {
      _logger.e('Failed to fetch history report', error: e);
      return _handleError(e);
    } catch (e) {
      _logger.e('History report unexpected error', error: e);
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Download Driver History Report as Excel
  Future<Map<String, dynamic>> downloadDriverHistoryReport({
    required String startDate,
    required String endDate,
    required String savePath,
  }) async {
    try {
      final queryParams = {
        'start_date': startDate,
        'end_date': endDate,
        'format': 'excel',
      };

      _logger.i('Downloading history report to: $savePath');
      final response = await _dio.download(
        ApiEndpoints.driverHistoryReport,
        savePath,
        queryParameters: queryParams,
      );

      return {'success': true};
    } on DioException catch (e) {
      _logger.e('Failed to download history report', error: e);
      return _handleError(e);
    } catch (e) {
      _logger.e('History report download error', error: e);
      return {'success': false, 'error': e.toString()};
    }
  }

  Map<String, dynamic> _handleError(DioException e) {
    String error = 'Network error';
    String? errorCode;

    final status = e.response?.statusCode;

    // Map status codes to fallback messages (used only when the backend sends
    // no body — e.g. pure network errors, timeouts, proxy errors).
    // When the backend DOES send a JSON body with a 'message' field, that
    // overrides these strings below.
    if (status == 401) {
      error = 'Session expired. Please log in again.';
      errorCode = 'UNAUTHORIZED';
    } else if (status == 403) {
      error = 'Access denied. Session may have expired.';
      errorCode = 'FORBIDDEN';
    } else if (status == 404) {
      error = 'Resource not found.';
      errorCode = 'NOT_FOUND';
    } else if (status == 409) {
      // Conflict — stop event already recorded for this booking.
      // The backend message (if present) will override 'error' below.
      error = 'A conflicting stop event already exists.';
      errorCode = 'STOP_EVENT_CONFLICT';
    } else if (status != null && status >= 500) {
      error = 'Server error. Please try again later.';
      errorCode = 'SERVER_ERROR';
    } else if (e.type == DioExceptionType.connectionTimeout ||
               e.type == DioExceptionType.receiveTimeout ||
               e.type == DioExceptionType.sendTimeout) {
      error = 'Connection timed out. Please check your network.';
      errorCode = 'TIMEOUT';
    } else if (e.type == DioExceptionType.connectionError) {
      error = 'No internet connection.';
      errorCode = 'NETWORK_ERROR';
    }

    // Try to override with server's own message if available
    if (e.response?.data != null) {
      final data = e.response?.data;
      if (data is Map) {
         if (data.containsKey('detail')) {
           final detail = data['detail'];
           if (detail is List) {
             error = detail.map((e) => "${e['loc']?.last}: ${e['msg']}").join(', ');
           } else if (detail is Map) {
             error = detail['message'] ?? error;
           } else if (detail is String && detail.isNotEmpty) {
             error = detail;
           }
         } else {
           final msg = data['message'] ?? data['error'];
           if (msg is String && msg.isNotEmpty) error = msg;
         }
         errorCode = data['code'] ?? data['error_code'] ?? errorCode;
      }
    }
    return {'success': false, 'error': error, 'errorCode': errorCode};
  }
}
